### PR TITLE
freebsd: various new packages to support a NixBSD system

### PR DESCRIPTION
**Pull request by rhelmot**
## Description of changes

61 of them,1 small package diff, 1 large package diff (sys), and 1 commit modifying lib and mkDerivation. This lets us boot NixBSD!

nixbsd neofetch

There are a few TODOs left in the code but I think they're fine for followup PRs.

## Things done

- Built on platform(s)
  - [ ] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
  - [x] x86_64-freebsd (cross from linux)
- For non-Linux: Is sandboxing enabled in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
  - [ ] `sandbox = relaxed`
  - [ ] `sandbox = true`
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [x] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [24.11 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2411.section.md) (or backporting [23.11](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) and [24.05](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2405.section.md) Release notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

---

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

### CODE REVIEW

File: pkgs/os-specific/bsd/freebsd/pkgs/drm-kmod/package.nix
Comment (JohnRTitor) [inline pkgs/os-specific/bsd/freebsd/pkgs/drm-kmod/package.nix:54]: CI errors suggests that `nixfmt` fails. Could you format with `nixfmt-rfc-style` instead?

Comment (JohnRTitor) [reply pkgs/os-specific/bsd/freebsd/pkgs/drm-kmod/package.nix:54]:
```nix
Checking paths: pkgs/os-specific/bsd
pkgs/os-specific/bsd/freebsd/pkgs/drm-kmod/package.nix: not formatted
Error: nixfmt failed.
Error: Process completed with exit code 1.
```

https://github.com/NixOS/nixpkgs/pull/320475/commits/bdee1e245820428f754be393d07562e358f6dd80

Specifically this causes the error, I wonder why.

Comment (paparodeo) [reply pkgs/os-specific/bsd/freebsd/pkgs/drm-kmod/package.nix:54]: > I am using that. Which rev of nixpkgs is new enough?

not sure if this is the answer to your question but the nixfmt used by the workflow is
https://github.com/NixOS/nixpkgs/blob/a9b494b6bcc31b6df8459b307448103412270006/.github/workflows/check-nix-format.yml#L28-L30

```
NIX_PATH="nixpkgs=https://github.com/NixOS/nixpkgs/archive/4b455dc2048f73a79eb3713f342369ff58f93e0b.tar.gz" nix-env -f '<nixpkgs>' -iAP nixfmt-rfc-style
```

Comment (rhelmot) [reply pkgs/os-specific/bsd/freebsd/pkgs/drm-kmod/package.nix:53]: I switched to that exact rev and it seems CI is happy now.

File: pkgs/os-specific/bsd/freebsd/lib/default.nix
Comment (AndersonTorres) [inline pkgs/os-specific/bsd/freebsd/lib/default.nix:23]: This is too long. Can it be at least documented?

File: pkgs/os-specific/bsd/freebsd/pkgs/drm-kmod-firmware.nix
Comment (AndersonTorres) [inline pkgs/os-specific/bsd/freebsd/pkgs/drm-kmod-firmware.nix:50]: This is ambiguous.

What does it mean?
a) does not require modification
b) prohibits any  modification

File: pkgs/os-specific/bsd/freebsd/pkgs/getty.nix
Comment (Ericson2314) [inline pkgs/os-specific/bsd/freebsd/pkgs/getty.nix:13]: Maybe this wrapping stuff should be left for later? Not really sure what it is yet.

File: pkgs/os-specific/bsd/freebsd/pkgs/rc.nix
Comment (Ericson2314) [inline pkgs/os-specific/bsd/freebsd/pkgs/rc.nix:29]: hmm?

File: pkgs/os-specific/bsd/freebsd/pkgs/ping.nix
Comment (Ericson2314) [inline pkgs/os-specific/bsd/freebsd/pkgs/ping.nix:17]: What's this?

File: pkgs/os-specific/bsd/freebsd/pkgs/zfs.nix
Comment (Ericson2314) [inline pkgs/os-specific/bsd/freebsd/pkgs/zfs.nix:30]: :eyes:

File: pkgs/os-specific/bsd/freebsd/lib/default.nix
Comment (Ericson2314) [reply pkgs/os-specific/bsd/freebsd/lib/default.nix:23]: It does need documentation, but do note it is just being moved. I would be OK with that happening in a later PR.

**Review by Ericson2314 — COMMENTED** [review]
I went through each commit, but I don't wanna nit-pick this to death :)

File: pkgs/os-specific/bsd/freebsd/pkgs/ping.nix
Comment (rhelmot) [reply pkgs/os-specific/bsd/freebsd/pkgs/ping.nix:17]: looks like it's removing a suid bit

File: pkgs/os-specific/bsd/freebsd/pkgs/getty.nix
Comment (rhelmot) [reply pkgs/os-specific/bsd/freebsd/pkgs/getty.nix:13]: What's wrong with it? we are in fact doing suid binary wrappers on nixbsd and they work fine. Would you rather it default to off? What's the practical use of running a non-suid login binary?

File: pkgs/os-specific/bsd/freebsd/pkgs/drm-kmod-firmware.nix
Comment (rhelmot) [reply pkgs/os-specific/bsd/freebsd/pkgs/drm-kmod-firmware.nix]: I took a look at the actual license file and it seems that it's the latter. I updated the verbage to be more explicit.

File: pkgs/os-specific/bsd/freebsd/pkgs/rc.nix
Comment (rhelmot) [reply pkgs/os-specific/bsd/freebsd/pkgs/rc.nix]: I took another dive and figured it out. All better!

File: pkgs/os-specific/bsd/freebsd/lib/default.nix
Comment (rhelmot) [reply pkgs/os-specific/bsd/freebsd/lib/default.nix:27]: I added an overview.

File: pkgs/os-specific/bsd/freebsd/pkgs/drm-kmod-firmware.nix
Comment (AndersonTorres) [inline pkgs/os-specific/bsd/freebsd/pkgs/drm-kmod-firmware.nix:52]:
```suggestion
  meta = {
    description = "GPU firmware for FreeBSD drm-kmod";
    platforms = lib.platforms.freebsd;
    license =
      lib.optional withAmd lib.licenses.unfreeRedistributableFirmware
      # Intel license prohibits modification. this will wrap firmware files in an ELF
      ++ lib.optional withIntel lib.licenses.unfree;
    sourceProvenance = [ lib.sourceTypes.binaryFirmware ];
```

File: pkgs/os-specific/bsd/freebsd/pkgs/drm-kmod/package.nix
Comment (AndersonTorres) [inline pkgs/os-specific/bsd/freebsd/pkgs/drm-kmod/package.nix:45]:
```suggestion
  meta = {
```

File: pkgs/os-specific/bsd/freebsd/pkgs/drm-kmod/package.nix
Comment (AndersonTorres) [inline pkgs/os-specific/bsd/freebsd/pkgs/drm-kmod/package.nix:48]:
```suggestion
    license = with lib.licenses; [
```

File: pkgs/os-specific/bsd/freebsd/pkgs/libzfs.nix
Comment (AndersonTorres) [inline pkgs/os-specific/bsd/freebsd/pkgs/libzfs.nix:84]:
```suggestion
  meta = {
    platforms = lib.platforms.freebsd;
    license = with lib.licenses; [ cddl ];
  };
```

File: pkgs/os-specific/bsd/freebsd/pkgs/syslogd.nix
Comment (AndersonTorres) [inline pkgs/os-specific/bsd/freebsd/pkgs/syslogd.nix:22]:
```suggestion
  meta = {
    description = "FreeBSD syslog daemon";
    maintainers = with lib.maintainers; [ artemist ];
    platforms = lib.platforms.freebsd;
    license = lib.licenses.bsd2;
  };
```

File: pkgs/os-specific/bsd/freebsd/pkgs/zfs.nix
Comment (AndersonTorres) [inline pkgs/os-specific/bsd/freebsd/pkgs/zfs.nix:48]:
```suggestion
  meta = {
    platforms = lib.platforms.freebsd;
    license = with lib.licenses; [
      cddl
      bsd2
    ];
  };
```

**Review by AndersonTorres — CHANGES_REQUESTED** [review]

File: pkgs/os-specific/bsd/freebsd/pkgs/getty.nix
Comment (Ericson2314) [reply pkgs/os-specific/bsd/freebsd/pkgs/getty.nix:13]: I just meant that `/run/wrappers` isn't a Nixpkgs thing, so it's arguably a minor layer violation I think?

Comment (rhelmot) [reply pkgs/os-specific/bsd/freebsd/pkgs/getty.nix]: I changed the parameter to be the wrappedLogin path itself, null if unused, as we discussed on matrix. You should go complain to the other package maintainers for their packages that hardcode wrapper paths!

**Review by JohnRTitor — APPROVED** [review]
Let's not hold this any further, subsequent improvements and bugfixes can come in a later PRs.

**Pull request merged**
